### ISSUE TITLE
Adapter recyclerview native ads

### ISSUE THREAD

**Issue by byJS93** [AppLovin/AppLovin-MAX-SDK-Android#151]
Hello, I have just had the ads in mopub since I had a good sdk for mediation.
Now that I see that they close it I am moving to this one but I see it with many failures.
I'm trying to implement native ads in a recyclerview and mopub built it into its own sdk for us but I've read the applovin documentation and I don't see a way to combine them.
In mopub I did it this way:

```
ViewBinder viewBinder = new ViewBinder.Builder(R.layout.ads_native)
                    .mainImageId(R.id.native_ad_main_image)
                    .iconImageId(R.id.native_ad_icon_image)
                    .titleId(R.id.native_ad_title)
                    .textId(R.id.native_ad_text)
                    .privacyInformationIconImageId(R.id.native_ad_privacy_information_icon_image)
                    .sponsoredTextId(R.id.native_sponsored_text_view)
                    .build();
MoPubStaticNativeAdRenderer adRenderer = new MoPubStaticNativeAdRenderer(viewBinder);
myMoPubAdapter = new MoPubRecyclerAdapter(this, imageadapterlistas, adPositioning);
myMoPubAdapter.registerAdRenderer(adRenderer);
listview.setAdapter(myMoPubAdapter);
myMoPubAdapter.loadAds(ID_Nativo);
```

Do you have a similar form in your sdk?
In mopub there was also a way in the announcement panel in which you decided every few elements of the list an announcement would come out and I see that in your panel you don't have that system either.

**Comment by chauduyphanvu** [comment]
@byJS93 - hello, and welcome to MAX. We are currently working on supporting this feature, and it should be ready around mid-March. Stay tuned!

Vu Chau
AppLovin Team

**Comment by byJS93** [comment]
> @byJS93 - hola y bienvenido a MAX. Actualmente estamos trabajando para admitir esta función y debería estar lista a mediados de marzo. ¡Manténganse al tanto!
> 
> Equipo Vu Chau AppLovin

I'll be waiting, thank you very much

**Comment by Luklek** [comment]
> 

But how are we supposed to start implementing MAX SDK in the middle of March, make sure everything working fine, upload to Google Play, pass review time, let every user update the app... all of this in less than 2 weeks before MoPub shut down?

**Comment by Luklek** [comment]
> @Luklek - we are working as hard as we can to support this feature and all the other ones. I have bumped the priority for this feature internally, and will get back to you when there's an ETA update. Thank you.

@chauduyphanvu nothing against you and the team effort, you have always been super helpful on MoPub github for years.
Just venting in hope that Applovin will extend the deadline to close Mopub...

**Comment by joeykrim** [comment]
For those who are brand new to Applovin MAX and feel 2 weeks window is too small and will result in lost revenue, can anybody share guidance on easiest path to building this functionality?
Take the old MoPub RecycleView code and start modifying every MoPub call to Applovin MAX call?
Even a temporary workaround is enough until there's formal support.

Depending how many other issues I have converting from MoPub to Applovin, this might be the most important so I'll try to update this comment with my progress.

**Comment by chauduyphanvu** [comment]
@joeykrim - one approach would be to write your own list adapter that injects loaded ads at the desired indices in your feed. Start by loading a couple of ads and holding them off of the view hierarchy. Then you can add your ads as child views to the ListView/RecyclerView as the user scrolls. That's the essence of the MoPub ad placer.

Here's the code for reference: https://github.com/mopub/mopub-android-sdk/blob/master/mopub-sdk/mopub-sdk-native-static/src/main/java/com/mopub/nativeads/MoPubStreamAdPlacer.java.

**Comment by Luklek** [comment]
@chauduyphanvu are there any news on ETA for the adapter?

**Comment by chauduyphanvu** [comment]
@Luklek - thanks for checking in. There are currently no update, but I'll keep you posted.

**Comment by tfkci** [comment]
any news ?

**Comment by chauduyphanvu** [comment]
Hi all, the feature is ready and undergoing testing. We plan to ship it for iOS first in ~2 weeks, and then for Android in about a month.

Thanks for your patience as we finalize the feature.

**Comment by laurencedawson** [comment]
Is this feature still not available? Mopub is about to close down...

**Comment by nihp** [comment]
Any updates on this?

**Comment by nihp** [comment]
I am also using the Mopub code for the feeds in my app. Any updates or any inputs to get pass over this will be helpful.

```
ViewBinder viewBinder = new ViewBinder.Builder(R.layout.native_ad_mopub)
                   .mainImageId(R.id.ad_cover)
                   .iconImageId(R.id.brand_logo)
                   .titleId(R.id.ad_title)
                   .textId(R.id.ad_description)
                   .privacyInformationIconImageId(R.id.native_privacy_information_icon_image)
                   .addExtra(context.getResources().getString(R.string.sponsored), R.id.description)

staticNativeAdRenderer = new MoPubStaticNativeAdRenderer(viewBinder);

moPubNative = new MoPubNative((Activity) context, key, new MoPubNative.MoPubNativeNetworkListener() {
})

       moPubNative.registerAdRenderer(staticNativeAdRenderer);
      final EnumSet<RequestParameters.NativeAdAsset> desiredAssets = EnumSet.of(
               RequestParameters.NativeAdAsset.TITLE,
               RequestParameters.NativeAdAsset.TEXT,
              RequestParameters.NativeAdAsset.ICON_IMAGE,
               RequestParameters.NativeAdAsset.MAIN_IMAGE,
               RequestParameters.NativeAdAsset.CALL_TO_ACTION_TEXT);

       RequestParameters requestParameters = new RequestParameters.Builder()
               .desiredAssets(desiredAssets)
               .build();

       moPubNative.makeRequest(requestParameters);
```

**Comment by nihp** [comment]
Any update on this?

**Comment by byJS93** [comment]
> Hola a todos, la función está lista y se está probando. Planeamos enviarlo para iOS primero en aproximadamente 2 semanas y luego para Android en aproximadamente un mes.
> 
> Gracias por su paciencia mientras finalizamos la función.

hello, almost 1 month has passed since you said you were testing it, is the function ready?

**Comment by tfkci** [comment]
Any updates about the issue?

**Comment by thomasmso** [comment]
We're close to a public release. If you'd like to be part of our internal beta, please send an email to thomas.so@applovin.com. Please include App / Play Store portfolios as well, thanks.

**Comment by nihp** [comment]
@thomasmso Any update on this. We are waiting for updated SDk with applovin above Ad Placer API.

**Comment by thomasmso** [comment]
We have the doc sand beta SDK ready, I communicated it over here on how to get access

> We're close to a public release. If you'd like to be part of our internal beta, please send an email to [thomas.so@applovin.com](mailto:thomas.so@applovin.com). Please include App / Play Store portfolios as well, thanks.

**Comment by ritamsarmah** [comment]
Hi all,

We have officially released Ad Placer in our latest Android SDK (11.4.0), allowing you to easily insert native ads into your existing content streams using `RecyclerView`. Check out our documentation for more information and integration instructions: https://dash.applovin.com/documentation/mediation/android/getting-started/native-ad-placer
